sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateOrdering.scala

```diff
@@ -138,3 +138,32 @@ object GenerateOrdering extends CodeGenerator[Seq[SortOrder], Ordering[InternalR
     CodeGenerator.compile(code).generate(ctx.references.toArray).asInstanceOf[BaseOrdering]
   }
 }
+
+/**
+ * A lazily generate row ordering comparator.
+ */
+class LazilyGenerateOrdering(val ordering: Seq[SortOrder]) extends Ordering[InternalRow] {
+
+  def this(ordering: Seq[SortOrder], inputSchema: Seq[Attribute]) =
+    this(ordering.map(BindReferences.bindReference(_, inputSchema)))
+
+  @transient
+  lazy val generatedOrdering = GenerateOrdering.generate(ordering)
```

Comment:
Will this use of `lazy val` incur a performance penalty due to having to check whether it has been initialized on each call to `compare()`? If so, I wonder whether we could eagerly initialize the ordering in the constructor then re-initialize on executors as part of `readObject()`, e.g. `@transient private[this] var generatedOrdering = ...`